containerd-shim-spin/src/engine.rs

```diff
@@ -290,6 +348,52 @@ impl Engine for SpinEngine {
     fn can_handle(&self, _ctx: &impl RuntimeContext) -> Result<()> {
         Ok(())
     }
+
+    fn supported_layers_types() -> &'static [&'static str] {
+        &[
+            OCI_LAYER_MEDIA_TYPE_WASM,
+            OCI_LAYER_MEDIA_TYPE_DATA,
+            OCI_LAYER_MEDIA_TYPE_SPIN_CONFIG,
+        ]
+    }
+
+    fn precompile(&self, layers: &[WasmLayer]) -> Result<Vec<Option<Vec<u8>>>> {
+        // Runwasi expects layers to be returned in the same order, so wrap each layer in an option, setting non Wasm layers to None
+        let mut wasm_content_layers: Vec<Option<WasmLayer>> =
+            layers.iter().map(SpinEngine::is_wasm_content).collect();
+        for layer in &mut wasm_content_layers {
+            if let Some(wasm_layer) = layer.as_mut() {
+                log::info!(
+                    "Precompile called for wasm layer {:?}",
+                    wasm_layer.config.digest()
+                );
+                if self
+                    .wasmtime_engine
+                    .detect_precompiled(&wasm_layer.layer)
+                    .is_some()
+                {
+                    log::info!("Layer already precompiled {:?}", wasm_layer.config.digest());
+                    continue;
+                }
+                let component = spin_componentize::componentize_if_necessary(&wasm_layer.layer)?;
+                let precompiled = self.wasmtime_engine.precompile_component(&component)?;
+                wasm_layer.layer = precompiled;
+            }
+        }
+        let precompiled_layers: Vec<Option<Vec<u8>>> = wasm_content_layers
+            .into_iter()
+            .map(|l| l.map(|l| l.layer))
+            .collect();
+        Ok(precompiled_layers)
+    }
```

Comment:
refactor:
```rust
layers.iter().map(|layer| {
        match SpinEngine::is_wasm_content(layer) {
            Some(mut wasm_layer) => {
                ...
             }
             None => Ok(None),
        }
    }).collect()
```

Reply:
this is much cleaner. Thank you @Mossaka